nav2_route/src/plugins/graph_file_loaders/geojson_graph_file_loader.cpp

```diff
@@ -0,0 +1,235 @@
+    return false;
+  }
+
+  if (edges.empty()) {
+    RCLCPP_ERROR(logger_, "No edges were found in %s", filepath.c_str());
+    return false;
+  }
+
+  graph.resize(nodes.size());
+  addNodesToGraph(graph, graph_to_id_map, nodes);
+  addEdgesToGraph(graph, graph_to_id_map, edges);
+
+  return true;
+}
+
+bool GeoJsonGraphFileLoader::fileExists(const std::string & filepath)
+{
+  return std::filesystem::exists(filepath);
+}
+
+
+void GeoJsonGraphFileLoader::getGraphElements(
+  const Json & features, std::vector<Json> & nodes, std::vector<Json> & edges)
+{
+  for (const auto & feature : features) {
+    if (feature["geometry"]["type"] == "Point") {
+      nodes.emplace_back(feature);
+    } else if (feature["geometry"]["type"] == "MultiLineString") {
+      edges.emplace_back(feature);
+    }
+  }
+}
+
+void GeoJsonGraphFileLoader::addNodesToGraph(
+  Graph & graph, GraphToIDMap & graph_to_id_map, std::vector<Json> & nodes)
+{
+  int idx = 0;
+  for (const auto & node : nodes) {
+    const auto properties = node["properties"];
+    graph[idx].nodeid = properties["id"];
+    graph_to_id_map[graph[idx].nodeid] = idx;
+    graph[idx].coords = getCoordinates(node);
+    graph[idx].operations = getOperations(properties);
+    graph[idx].metadata = getMetaData(properties);
+    idx++;
+  }
+}
+
+void GeoJsonGraphFileLoader::addEdgesToGraph(
+  Graph & graph, GraphToIDMap & graph_to_id_map, std::vector<Json> & edges)
+{
+  for (const auto & edge : edges) {
+    // Required data
+    const auto properties = edge["properties"];
+    unsigned int id = properties["id"];
+    unsigned int start_id = properties["startid"];
+    unsigned int end_id = properties["endid"];
+
+    if (graph_to_id_map.find(start_id) == graph_to_id_map.end()) {
+      RCLCPP_ERROR(logger_, "Start id %u does not exist for edge id %u", start_id, id);
+      throw nav2_core::NoValidGraph("Start id does not exist");
+    }
+
+    if (graph_to_id_map.find(end_id) == graph_to_id_map.end()) {
+      RCLCPP_ERROR(logger_, "End id of %u does not exist for edge id %u", end_id, id);
+      throw nav2_core::NoValidGraph("End id does not exist");
+    }
+
+    EdgeCost edge_cost = getEdgeCost(properties);
+    Operations operations = getOperations(properties);
+    Metadata metadata = getMetaData(properties);
+
+    graph[graph_to_id_map[start_id]].addEdge(
+      edge_cost, &graph[graph_to_id_map[end_id]], id,
+      metadata, operations);
+  }
+}
+
+Coordinates GeoJsonGraphFileLoader::getCoordinates(const Json & node)
+{
+  Coordinates coords;
+  const auto & properties = node["properties"];
+  if (properties.contains("frame")) {
+    coords.frame_id = properties["frame"];
+  }
+
+  const auto & coordinates = node["geometry"]["coordinates"];
+  coords.x = coordinates[0];
+  coords.y = coordinates[1];
+
+  return coords;
+}
+
+Metadata GeoJsonGraphFileLoader::getMetaData(const Json & properties, const std::string & key)
```

Comment:
The key is never set to anything else. Why not just have `properties.contains("metadata")` like we have hardcoded for other things like frames / cost / etc?

Reply:
It is needed for the recursive calls.

Reply:
Oh I see, I missed one of the calls. Got it!